tensorflow/python/ops/array_ops.py

```diff
@@ -1757,6 +1757,176 @@ def _DepthToSpaceShape(op):
+    on_value = 0.0
+    off_value = 3.0
+    axis = 0
+  ```
+
+  Then output is `[3 x 4]`:
+
+  ```
+    output =
+    [0.0 3.0 3.0 3.0]
+    [3.0 3.0 3.0 0.0]
+    [3.0 3.0 3.0 3.0]
+    [3.0 0.0 3.0 3.0]
+  //  ^                one_hot(0)
+  //      ^            one_hot(2)
+  //          ^        one_hot(-1)
+  //              ^    one_hot(1)
+  ```
+
+  Suppose that
+
+  ```
+    indices = [[0, 2], [1, -1]]
+    depth = 3
+    on_value = 1.0
+    off_value = 0.0
+    axis = -1
+  ```
+
+  Then output is `[2 x 2 x 3]`:
+
+  ```
+    output =
+    [
+      [1.0, 0.0, 0.0]  // one_hot(0)
+      [0.0, 0.0, 1.0]  // one_hot(2)
+    ][
+      [0.0, 1.0, 0.0]  // one_hot(1)
+      [0.0, 0.0, 0.0]  // one_hot(-1)
+    ]
+  ```
+
+  Args:
+    indices: A `Tensor` of indices.
+    depth: A scalar defining the depth of the one hot dimension.
+    on_value: A scalar defining the value to fill in output when `indices[j]
+      = i`. (default: 1)
+    off_value: A scalar defining the value to fill in output when `indices[j]
+      != i`. (default: 0)
+    axis: The axis to fill (default: -1, a new inner-most axis).
+    dtype: The data type of the output tensor.
+
+  Returns:
+    output: The one-hot tensor.
+
+  Raises:
+    TypeError: If on_value or off_value do not match dtype.
+    TypeError: If dtype is `tf.string`
+  """
+  # Check for bad dtype specification
+  if dtype == dtypes.string:
+    raise TypeError("dtype must be a numeric type")
+
+  with ops.op_scope([indices, depth, on_value, off_value,
+            axis, dtype], name, "one_hot") as name:
+    on_exists = on_value is not None
+    off_exists = off_value is not None
+    dtype_exists = dtype is not None
+
+    # If only one of on_value and off_value is provided, and dtype is not given,
+    # Assign dtype to the type of provided value
+    if not dtype_exists:
+      if on_exists and not off_exists:
+        dtype = ops.convert_to_tensor(on_value).dtype.base_dtype
+        dtype_exists = True
+      elif off_exists and not on_exists:
+        dtype = ops.convert_to_tensor(off_value).dtype.base_dtype
+        dtype_exists = True
+
+    # Create on_value tensor
+    if on_exists:
+      on_value = ops.convert_to_tensor(on_value, name="on_value")
+      if dtype_exists and on_value.dtype.base_dtype != dtype:
+        raise TypeError("The type of on_value must match dtype")
+    elif dtype_exists:
+      on_value = constant(1, dtype, name="on_value")
+    else:
+      # Neither on_value nor dtype are provided
+      on_value = constant(1, dtypes.float32, name="on_value")
+
+    # Create off_value tensor 
+    if off_exists:
+      off_value = ops.convert_to_tensor(off_value, name="off_value")
+      if dtype_exists and off_value.dtype.base_dtype != dtype:
+        raise TypeError("The type of off_value must match dtype")
+    elif dtype_exists:
+      off_value = constant(0, dtype, name="off_value")
+    else:
+      off_value = constant(0, dtypes.float32, name="off_value")
+    
```

Comment:
Here's some simplified logic for all the exists cases, once the defaults are fixed to be their actual values rather than `None`.

    if dtype is None:
      try:
        dtype = on_value.dtype
      except AttributeError:
        try:
          dtype = off_value.dtype
        except AttributeError:
          pass
    on_value = ops.convert_to_tensor(on_value, dtype=dtype, name="on_value")
    off_value = ops.convert_to_tensor(on_value, dtype=dtype, name="off_value")